listenbrainz/webserver/static/js/src/RecentListens.tsx

```diff
@@ -586,6 +595,194 @@ export default class RecentListens extends React.Component<
+    }
+  };
+
+  handleClickNewerRecommendations = () => {
+    const { listens } = this.props;
+    const { offset } = this.state;
+
+    if (
+      (offset || offset === 0) &&
+      offset >= this.expectedRecommendationsPerPage
+    ) {
+      const newOffset = offset - this.expectedRecommendationsPerPage;
+      this.setState({ loading: true });
+      this.setState(
+        {
+          offset: newOffset,
+          listens:
+            listens?.slice(
+              newOffset,
+              newOffset + this.expectedRecommendationsPerPage
+            ) || [],
+        },
+        this.afterRecommendationsDisplay
+      );
+      window.history.pushState(null, "", "");
+    }
+  };
+
+  handleClickOlderRecommendations = () => {
+    const { listens } = this.props;
+    const { offset, recommendationLength } = this.state;
+
+    const condition = recommendationLength
+      ? recommendationLength - this.expectedRecommendationsPerPage
+      : 0;
+    if ((offset || offset === 0) && condition && offset < condition) {
+      const newOffset = offset + this.expectedRecommendationsPerPage;
+      this.setState({ loading: true });
+      this.setState(
+        {
+          offset: newOffset,
+          listens:
+            listens?.slice(
+              newOffset,
+              newOffset + this.expectedRecommendationsPerPage
+            ) || [],
+        },
+        this.afterRecommendationsDisplay
+      );
+      window.history.pushState(null, "", "");
+    }
+  };
+
+  handleClickOldestRecommendations = () => {
+    const { listens } = this.props;
+    const { recommendationLength, offset } = this.state;
+
+    const condition = recommendationLength
+      ? recommendationLength - this.expectedRecommendationsPerPage
+      : 0;
+    if ((offset || offset === 0) && condition && offset < condition) {
+      this.setState({ loading: true });
+
+      const newOffset =
+        (Math.ceil(
+          recommendationLength
+            ? recommendationLength / this.expectedRecommendationsPerPage
+            : 0
+        ) -
+          1) *
+        this.expectedRecommendationsPerPage;
+
+      this.setState(
+        {
+          listens: listens?.slice(newOffset, recommendationLength) || [],
+          offset: newOffset,
+        },
+        this.afterRecommendationsDisplay
+      );
+      window.history.pushState(null, "", "");
+    }
+  };
+
+  handleRecommendationChange = () => {
+    const { offset, recommendationLength } = this.state;
+
+    return (
+      <ul className="pager" style={{ display: "flex" }}>
+        <li
+          className={`previous ${
+            (offset || offset === 0) &&
+            offset < this.expectedRecommendationsPerPage
+              ? "disabled"
+              : ""
+          }`}
+        >
+          <a
+            role="button"
+            onClick={this.handleClickNewestRecommendations}
+            onKeyDown={(e) => {
+              if (e.key === "Enter") this.handleClickNewestRecommendations();
```

Comment:
Rather than creating 4 different handlers for each button we can pass the offset to a single function and remove lot of redundant code.